src/sensors/d200.cpp

```diff
@@ -0,0 +1,103 @@
+#include "d200.hpp"
+
+D200LD14P::D200LD14P() {
+  state.buffer_idx = 0;
+  LIDAR_SERIAL.begin(LIDAR_BAUD);
+  start();
+}
+
+LidarDataPacket D200LD14P::get_latest_packet() {
+  return latest_packet;
+}
+
+uint8_t D200LD14P::calc_checksum(uint8_t *buf, int len) {
+  uint8_t crc8 = 0;
+  for (int i = 0; i < len; i++) {
+    crc8 = CRC_TABLE[(crc8 ^ *buf++) & 0xff];
+  }
+  return crc8;
+}
+
+void D200LD14P::set_speed(uint16_t speed) {
+  uint8_t lsb = speed & 0xff;
+  uint8_t msb = (speed >> 8) & 0xff;
+
+  uint8_t cmd[CMD_LEN] = { 0x54, 0xa2, 0x04, lsb, msb, 0, 0, 0 };
+  cmd[CMD_LEN - 1] = calc_checksum(cmd, CMD_LEN - 1);
+
+  LIDAR_SERIAL.write(cmd, CMD_LEN);
+}
+
+void D200LD14P::start() {
+  LIDAR_SERIAL.write(START_CMD, CMD_LEN);
+}
+
+void D200LD14P::stop() {
+  LIDAR_SERIAL.write(STOP_CMD, CMD_LEN);
+}
+
+void D200LD14P::read() {
+  while (LIDAR_SERIAL.available()) {
```

Comment:
I really like how you handled this to support any length of serial buffer. However, two edge cases to fix/verify:

**Case 1:**
1. You have a partially filled buffer, with say, 20 bytes in it, from a previous read() call, and are awaiting the next 27 bytes.
2. Time passes such that LIDAR_SERIAL gets full and overwrites old data multiple times. 
    Now, the partially filled packet would be invalid to continue filling, but code doesn't know that yet.
3. Read() is called some time later. Maybe a new start byte is encountered, maybe not, but regardless your code currently will not reset the buffer index = 0. Instead, it will read the next 27 garbage bytes as valid data. The checksum might catch it, but this isn't guarenteed since measurement data section could be anything (unless I am misunderstanding the checksum standard).
This edge case looks broken to me. 

I think this could be fixed by checking if LIDAR_SERIAL.available() == MAX_SERIAL_BUFFER. If true, it is possible something was overwritten. So, throw out the in-progress packet (reset idx = 0) and begin checking for a new start character instead.
This solution comes at a cost of lost packets, but I think correctness is more important here since this shouldn't happen too often unless the main loop code is too slow.


**Case 2:**
LIDAR_SERIAL has more than PACKET_SIZE bytes, so two possibilities exist:
there is one complete packet and one incomplete packet, or two incomplete packets. 

I ran through possibilities in my head and it looks like your solution functions, but you should make sure this performs as expected (don't repeatedly drop packets, etc).

Reply:
I disagree with this.

Reply:
You disagree with the edge case existing?

Reply:
Code was restructured, this analysis no longer applies